types/react-query/index.d.ts

```diff
@@ -1,142 +1,420 @@
+): PaginatedQueryResult<TResult>;
+
+export function usePaginatedQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables = []>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: QueryFunctionWithVariables<TResult, TKey, TVariables>;
+    config?: QueryOptions<TResult>;
+}): PaginatedQueryResult<TResult>;
+
+// useInfiniteQuery
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, TKey, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: InfiniteQueryFunction<TResult, [TKey], TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
+
+export function useInfiniteQuery<TResult, TKey extends string, TVariables extends AnyVariables, TMoreVariable>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, [TKey], TVariables, TMoreVariable>,
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>,
+): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+export function useInfiniteQuery<
+    TResult,
+    TKey extends AnyQueryKey,
+    TMoreVariable,
+    TVariables extends AnyVariables = []
+>({
+    queryKey,
+    variables,
+    queryFn,
+    config,
+}: {
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined);
+    variables?: TVariables;
+    queryFn: InfiniteQueryFunctionWithVariables<TResult, TKey, TVariables, TMoreVariable>;
+    config?: InfiniteQueryOptions<TResult, TMoreVariable>;
+}): InfiniteQueryResult<TResult, TMoreVariable>;
 
-export interface QueryOptions<TResult> {
+export type QueryKeyPart = string | object | boolean | number | null | readonly QueryKeyPart[] | null | undefined;
+export type AnyQueryKey = readonly [string, ...QueryKeyPart[]]; // this forces the key to be inferred as a tuple
```

Comment:
I meant to comment 👆

Reply:
ah, ok, it might be a valid concern.
It has to be a different overload though. Because in this case, we have to infer the type of the key as `[T]` where `T extends string`. I will make an update soon.

Reply:
addressed here #43568